### PR TITLE
backend/enhancement : Added Scheduled bookings in redis

### PR DESCRIPTION
**Pull request by dushyant-2002**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added integration tests for my changes where possible
- [ ] I tested the changes using integration tests
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/dev/migrations-read-only/dynamic-offer-driver-app/transporter_config.sql [hunk 1]
```diff
@@ -317,4 +317,9 @@ ALTER TABLE atlas_driver_offer_bpp.transporter_config ADD COLUMN otp_ride_start_
 
 ------- SQL updates -------
 
-ALTER TABLE atlas_driver_offer_bpp.transporter_config ADD COLUMN disable_list_scheduled_booking_api boolean  default false;
\ No newline at end of file
+ALTER TABLE atlas_driver_offer_bpp.transporter_config ADD COLUMN disable_list_scheduled_booking_api boolean  default false;
```
Comment (0utkarsh) [inline Backend/dev/migrations-read-only/dynamic-offer-driver-app/transporter_config.sql:320]: Make this true now?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Booking.hs [hunk 2]
```diff
@@ -85,3 +95,57 @@ cancelBooking booking mbDriver transporter = do
...
+  let (year, month, day) = toGregorian $ utctDay time
+      date = printf "%04d-%02d-%02d" year month day
+      cityId = unpack mocId.getId
+   in (pack $ "ScheduledBookings:" <> cityId <> ":" <> date)
+
+calculateSortedSetScore :: UTCTime -> Double
+calculateSortedSetScore time = realToFrac (utcTimeToPOSIXSeconds time)
+
+secondsRemainingInDay :: UTCTime -> Int
+secondsRemainingInDay currentTime =
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Booking.hs:139]: `ceiling $ 86400 - utctDayTime currentTime` only this is enough

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Cancel.hs [hunk 3]
```diff
@@ -153,6 +155,15 @@ reAllocateBookingIfPossible isValueAddNP userReallocationEnabled merchant bookin
...
       QRB.createBooking newBooking
+      when newBooking.isScheduled $ do
+        let score = ceiling $ calculateSortedSetScore newBooking.startTime
+            expirationSeconds = secondsRemainingInDay newBooking.startTime
+            vehicleVariant = castServiceTierToVariant newBooking.vehicleServiceTier
+        let redisKey = createRedisKey newBooking.startTime newBooking.merchantOperatingCityId newBooking.tripCategory vehicleVariant
+            redisKeyForHset = createRedisKeyForHset newBooking.startTime newBooking.merchantOperatingCityId
+            member = createMember newBooking
+        void $ Redis.zAddExp redisKey member score expirationSeconds
+        void $ Redis.hSetExp redisKeyForHset newBooking.id.getId newBooking expirationSeconds
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Cancel.hs:166]: This code is repeated during booking creation. Can be refactored

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs [hunk 4]
```diff
@@ -2239,6 +2273,42 @@ listScheduledBookings (personId, _, cityId) mbLimit mbOffset mbFromDay mbToDay m
...
+          endScore = calculateSortedSetScore $ addUTCTime (3600 * 24) dayStartTime
+
+      res <- mapM (\key -> Redis.zRangeByScoreByCount key startScore endScore offset limit) redisKeys
+
+      let parsedRes = mapMaybe (parseMember . decodeUtf8) (concat res)
+          nearbyBookings = filterNearbyBookings now dLoc vehicle.variant transporterConfig.avgSpeedOfVehicle parsedRes possibleServiceTierTypes
+
+      if not $ null nearbyBookings
+        then Redis.hmGet redisKeyForHset nearbyBookings
+        else pure []
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:2311]: Both these functions are very similar. Can you create one common function for these?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Init.hs [hunk 5]
```diff
@@ -126,12 +128,21 @@ handler merchantId req validatedReq = do
...
         triggerBookingCreatedEvent BookingEventData {booking = booking, personId = driverQuote.driverId, merchantId = transporter.id}
         QRB.createBooking booking
-
         QST.updateStatus DST.COMPLETED (searchTry.id)
         return (booking, Just driverQuote.driverName, Just driverQuote.driverId.getId)
       ValidatedQuote quote -> do
         booking <- buildBooking searchRequest quote quote.id.getId quote.tripCategory now (mbPaymentMethod <&> (.id)) paymentUrl Nothing req.initReqDetails
         QRB.createBooking booking
+        when booking.isScheduled $ do
+          let score = ceiling $ calculateSortedSetScore booking.startTime
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Init.hs:137]: Are we not including lat long in score calculation?

**Review by 0utkarsh — CHANGES_REQUESTED** [review]

**Review by arjuna92 — APPROVED** [review]

**Review by 0utkarsh — APPROVED** [review]

**Pull request merged**